### DISCUSSION TITLE
Application-wide audit of all dataportal calls (well, only the ones that modify data)

### DISCUSSION THREAD

**Discussion by vijaygill** [MarimerLLC/csla#2721] [Questions] [answered]
Hi,

I am required to implement logging of every datapoirtal call which leads to changing of data. Only changed properties need to be logged and a collection of those can then be saved as JSON in the logs along with the user identity.

All the DataPortal operations are server side, so logging is to be done at one place.

Thanks.

**Comment by rockfordlhotka** [comment, marked as answer]
Hi @vijaygill,

This should be possible with some work.

The server-side data portal has hooks where you can implement logging. The _Using CSLA 4_ book on the data portal covers these hooks up through CSLA 5. There are some changes in CSLA 6, though your options are similar. So it should be pretty easy to choose one of the options and implement logging.

Actually knowing which properties are changed is a little harder. The default implementation of the field manager doesn't maintain the original value retrieved from the database (due to size in memory and over the network). 

One way to know _for sure_ which properties have been changed is to maintain that original value so you can compare the current value to the original. The cost of this is memory and data over the wire.

The other way to know _within reason_ that properties have changed is to compare the current value in the object to the current value in the database. This has the cost of pulling the data from the db for comparison and isn't a perfect answer, b/c someone else could have changed the db in the meantime.

Another way to know _within reason_ is to store a hash code of the original value in the field manager, and then compare that hash to a hash of the current value to see if they are different (for your logging). Hash codes aren't guaranteed to be unique, so there's a non-zero chance of a changed property not being detected. You'd also be maintaining and transporting over the network an int (hash value) for each property, which is a small cost.

**Reply by vijaygill** [reply]
Thanks @rockfordlhotka for such a detailed answer. I take it that you mean Chapter 4 (Custom Data Portal Proxies) of that book is the partial solution to my requirement. I had gone though it but I needed some advice from you or more experienced (than me) user of CSLA.

I will further investigate on how to approach this "log changed fields only" approach.

Thank you, again for replying to my question.

**Reply by rockfordlhotka** [reply]
You shouldn't need to create a custom data portal or proxy. You can implement `IAuthorizeDataPortal`, `IDataPortalActivator`, or the override the `DataPortal_XYZ` meta methods (for invoke, invoke complete, exception) in your base classes.

The first thing that happens on the server in a data portal call is `IAuthorizeDataPortal`, which _is_ intended to support global authorization of data portal calls, but is also a place to do global logging.

The `IDataPortalActivator` is intended to allow you to use a DI container (this was before CSLA 5/6 where DI is directly supported) to create and clean up your DI container and various domain type instances. It can also be used for things like logging.

The virtual `DataPortal_XYZ` meta methods were originally intended specifically for things like logging, because they are running _inside_ your domain object instance(s), and so have full access to the object's state, metastate, etc.

In CSLA 6, thanks to @TheCakeMonster, there's now a rich way to build interceptors into the pipeline, where you can do all sort of things, including creating a global logging interceptor.

In my mind, I'd either overload the `DataPortal_XYZ` meta methods in your common base classes, or use the CSLA 6 interceptor concept. However, the _easiest_ place to put global logging is in an `IAuthorizeDataPortal` implementation.

**Reply by rockfordlhotka** [reply]
I should say that if `IAuthorizeDataPortal` rejects the call due to authorization rules, then nothing else runs on the server. So if you want to log _all calls_, even unauthorized attempts, then this is the only place you could log such a thing.

**Reply by vijaygill** [reply]
That's great! I am actually testing my code with CSLA 6 (pre-release), so I am interested in interceptors/custom activator too. I will have a look at csla/Samples/CustomActivator/ along with IAuthorizeDataPortal.

Thank you and other CSL:A contrinutors for all the good work.

**Reply by hurcane** [reply]
On the topic of tracking changed values, we have taken the approach of keeping track of a dictionary of changed values. The values in the dictionary are only set when the property is changed on on "old" object.

First, you need an interface to extend IBusinessBase and allow the logging code access to the original values.

```
public interface IChangeTracker : Csla.IBusinessBase
{
    Dictionary<string, object> ChangedValues { get; }
}
```
Then you need a subclass of BusinessBase to keep track of the original values as they are changed.

```
[Serializable]
public abstract class ChangeTracker<T> : Csla.BusinessBase<T>, IChangeTracker where T : ChangeTracker<T>
{
    MobileDictionary<string, object> _changedValues = new MobileDictionary<string, object>();
    public Dictionary<string, object> ChangedValues { get => _changedValues; }

    protected override void OnGetChildren(SerializationInfo info, MobileFormatter formatter)
    {
        SerializationInfo changeInfo = formatter.SerializeObject(_changedValues);
        info.AddChild(nameof(_changedValues), changeInfo.ReferenceId);
        base.OnGetChildren(info, formatter);
    }

    protected override void OnSetChildren(SerializationInfo info, MobileFormatter formatter)
    {
        SerializationInfo.ChildData changeData = info.Children[nameof(_changedValues)];
        _changedValues = (MobileDictionary<string, object>)formatter.GetObject(changeData.ReferenceId);
        base.OnSetChildren(info,
            formatter);
    }

    protected override void OnPropertyChanging(string propertyName)
    {
        // Changed values tracking only matters for existing objects.
        if (!this.IsNew)
        {
            // First time the property is changing, we need to store the original value.
            if (!_changedValues.ContainsKey(propertyName))
            {
                var field = FieldManager.GetRegisteredProperty(propertyName);
                _changedValues[propertyName] = FieldManager.GetFieldData(field).Value;
            }
        }
        base.OnPropertyChanging(propertyName);
    }

    protected override void PropertyHasChanged(IPropertyInfo property)
    {
        // Changed values tracking only matters for existing objects.
        if (!this.IsNew)
        {
            // Clear the field if the values match.
            if (_changedValues.ContainsKey(property.Name))
            {
                if (FieldManager.GetFieldData(property).Value.Equals(_changedValues[property.Name]))
                {
                    _changedValues.Remove(property.Name);
                }
            }
        }
        base.PropertyHasChanged(property);
    }

    protected override void MarkOld()
    {
        // When a class is being marked old. None of the properties are considered dirty any longer.
        _changedValues.Clear();
        base.MarkOld();
    }

    public override bool IsSelfDirty
    {
        get
        {
            if (this.IsNew || this.IsDeleted) return true;
            return this.ChangedValues.Count() > 0;
        }
    }
}
```

**Reply by rockfordlhotka** [reply]
iirc, FieldManager already keeps track of whether each property has been changed _at all_. It doesn't keep track of whether the value gets changed back to the original value, but each field (property) has its own IsDirty property. Using that might help you avoid all the extra tracking you've built?

**Reply by hurcane** [reply]
IsDirty is a one-way action in FieldManager. We wanted to support changing a value back to its original value and avoid writing changes to the database when there were no actual changes from the perspective of what was read from the database. In addition, we have an audit log requirement that records the old and new values, and the FieldManager does not keep track of the original value.

**Reply by rockfordlhotka** [reply]
Oh, I see what you are after.

FieldManager is interface-based with the intent that if you wanted to implement a different IFieldManager that did track the original value, you could do such a thing. Then your FieldManager.IsDirty could report based on comparing to the original value.

**Reply by hurcane** [reply]
We tried implementing the necessary interfaces, based on an old example from Jason Bock. However, that implementation carried a copy of every value. The sample used hashes for strings, but we needed the full original value. We found that implementation was pretty heavy from a memory perspective. This technique with the dictionary is much more lightweight.

**Reply by hurcane** [reply]
My main reason for posting that code was because the OP was making a request that we have had for a while now, and I thought it might be helpful when the need arises to log the original value.

**Reply by rockfordlhotka** [reply]
I like your approach for sure!

I still think there should be some way to create a custom FieldManager that accomplishes that goal without a high cost, though maybe I'm wrong. Something that only copies the original value on first change of the property?

It doesn't really matter, in that you have a good solution.

**Reply by vijaygill** [reply]
Thank you @hurcane and @rockfordlhotka for such invaluable discussion and insight into this.
